docs/middleware/callbacks/integration.md

```diff
@@ -0,0 +1,98 @@
+
+The callbacks middleware also **requires** a secondary application that will receive the callbacks to implement the [`ContractKeeper`](https://github.com/cosmos/ibc-go/blob/main/modules/apps/callbacks/types/expected_keepers.go#L11-L64). Since the wasm module does not yet support the callbacks middleware, we will use the mockContractKeeper module in the examples below. You should replace this with your own module that implements `ContractKeeper`.`
+
+### Transfer
+
+See below for an example of how to create an application stack using `transfer`, `29-fee`, and `callbacks`. Feel free to omit the `29-fee` middleware if you do not want to use it.
+The following `transferStack` is configured in `app/app.go` and added to the IBC `Router`.
+The in-line comments describe the execution flow of packets between the application stack and IBC core.
+
+```go
+// Create Transfer Stack
+// SendPacket, since it is originating from the application to core IBC:
+// transferKeeper.SendPacket -> callbacks.SendPacket -> fee.SendPacket -> channel.SendPacket
+
+// RecvPacket, message that originates from core IBC and goes down to app, the flow is the other way
+// channel.RecvPacket -> callbacks.OnRecvPacket -> fee.OnRecvPacket -> transfer.OnRecvPacket
+
+// transfer stack contains (from top to bottom):
+// - IBC Callbacks Middleware
+// - IBC Fee Middleware
+// - Transfer
+
+// create IBC module from bottom to top of stack
+var transferStack porttypes.IBCModule
+transferStack = transfer.NewIBCModule(app.TransferKeeper)
+transferStack = ibcfee.NewIBCMiddleware(transferStack, app.IBCFeeKeeper)
+transferStack = ibccallbacks.NewIBCMiddleware(transferStack, app.IBCFeeKeeper, app.MockContractKeeper, maxCallbackGas)
+// Since the callbacks middleware itself is an ics4wrapper, it needs to be passed to the transfer keeper
+app.TransferKeeper.WithICS4Wrapper(transferStack.(porttypes.Middleware))
+
+// Add transfer stack to IBC Router
+ibcRouter.AddRoute(ibctransfertypes.ModuleName, transferStack)
+```
+
+::: warning
+The usage of `WithICS4Wrapper` after `transferStack`'s configuration is critical! It allows the callbacks middleware to do `SendPacket` callbacks and asynchronous `ReceivePacket` callbacks. You must do this regardless of whether you are using the `29-fee` middleware or not.
+:::
+
+### Interchain Accounts Controller
+
+```go
+// Create Interchain Accounts Stack
+// SendPacket, since it is originating from the application to core IBC:
+// icaControllerKeeper.SendTx -> callbacks.SendPacket -> fee.SendPacket -> channel.SendPacket
+
+// initialize ICA module with mock module as the authentication module on the controller side
+var icaControllerStack porttypes.IBCModule
+icaControllerStack = ibcmock.NewIBCModule(&mockModule, ibcmock.NewIBCApp("", scopedICAMockKeeper))
```

Comment:
Is there an example/test of it being wired in this way? Or can I do this in callbacks' simapp and update the docs in another PR?

Reply:
```go
    var icaControllerStack porttypes.IBCModule
    icaControllerStack = icacontroller.NewIBCMiddleware(nil, app.ICAControllerKeeper)
    icaControllerStack = ibcfee.NewIBCMiddleware(icaControllerStack, app.IBCFeeKeeper)
    icaControllerStack = ibccallbacks.NewIBCMiddleware(icaControllerStack, app.IBCFeeKeeper, app.MockContractKeeper, maxCallbackGas)
    // Since the callbacks middleware itself is an ics4wrapper, it needs to be passed to the ica controller keeper
    app.ICAControllerKeeper.WithICS4Wrapper(icaControllerStack.(porttypes.Middleware))
```

You can simply provide the underlying app as nil to the controller module

As noted in the above comment, I think making callbacks the middle module would be nice, so:
```go
    var icaControllerStack porttypes.IBCModule
    icaControllerStack = icacontroller.NewIBCMiddleware(nil, app.ICAControllerKeeper)
    icaControllerStack = ibccallbacks.NewIBCMiddleware(icaControllerStack, app.IBCFeeKeeper, app.MockContractKeeper, maxCallbackGas)
    // Since the callbacks middleware itself is an ics4wrapper, it needs to be passed to the ica controller keeper
    app.ICAControllerKeeper.WithICS4Wrapper(icaControllerStack.(porttypes.Middleware))
    
    icaControllerStack = ibcfee.NewIBCMiddleware(icaControllerStack, app.IBCFeeKeeper)
```

Happy to see a followup pr for this

Reply:
I've made the underlying app `nil`, please review if the wiring looks correct @crodriguezvega 

```diff
// Create Interchain Accounts Stack
// SendPacket, since it is originating from the application to core IBC:
// icaControllerKeeper.SendTx -> callbacks.SendPacket -> fee.SendPacket -> channel.SendPacket

- // initialize ICA module with mock module as the authentication module on the controller side
var icaControllerStack porttypes.IBCModule
- icaControllerStack = ibcmock.NewIBCModule(&mockModule, ibcmock.NewIBCApp("", scopedICAMockKeeper))
- app.ICAAuthModule = icaControllerStack.(ibcmock.IBCModule)
- icaControllerStack = icacontroller.NewIBCMiddleware(icaControllerStack, app.ICAControllerKeeper)
+ icaControllerStack = icacontroller.NewIBCMiddleware(nil, app.ICAControllerKeeper)
icaControllerStack = ibcfee.NewIBCMiddleware(icaControllerStack, app.IBCFeeKeeper)
// maxCallbackGas is a hard-coded value that is passed to the callbacks middleware
icaControllerStack = ibccallbacks.NewIBCMiddleware(icaControllerStack, app.IBCFeeKeeper, app.MockContractKeeper, maxCallbackGas)
// Since the callbacks middleware itself is an ics4wrapper, it needs to be passed to the ica controller keeper
app.ICAControllerKeeper.WithICS4Wrapper(icaControllerStack.(porttypes.Middleware))

// RecvPacket, message that originates from core IBC and goes down to app, the flow is:
// channel.RecvPacket -> callbacks.OnRecvPacket -> fee.OnRecvPacket -> icaHost.OnRecvPacket

var icaHostStack porttypes.IBCModule
icaHostStack = icahost.NewIBCModule(app.ICAHostKeeper)
icaHostStack = ibcfee.NewIBCMiddleware(icaHostStack, app.IBCFeeKeeper)

- // Add host, controller & ica auth modules to IBC router
- ibcRouter.
- // the ICA Controller middleware needs to be explicitly added to the IBC Router because the
- // ICA controller module owns the port capability for ICA. The ICA authentication module
- // owns the channel capability.
+ // Add ICA host and controller to IBC router ibcRouter.
AddRoute(icacontrollertypes.SubModuleName, icaControllerStack).
AddRoute(icahosttypes.SubModuleName, icaHostStack).
- AddRoute(ibcmock.ModuleName+icacontrollertypes.SubModuleName, icaControllerStack) // ica with mock auth module stack route to ica (top level of middleware stack)
```